### ISSUE TITLE
Null checks for in-memory queries

### ISSUE THREAD

**Issue by rspeele** [ckimes89/graphql-net#30]
If, as currently, we translate:
```
  {
     user(id:1) { id, name, bestFriend { id, name } }
  }
```
To something like:
```
  db.Users.Select(u => new
    { id = u.Id
    , name = u.Name
    , bestFriend = new
      { id = u.BestFriend.Id
      , name = u.BestFriend.Name
      }
    })
```
This works in Entity Framework and probably most other ORMs but not in memory. If u.BestFriend is null, it will blow up with an NRE while executing the query. Inserting null checks can get really ugly and make the query expression incompatible with ORMs. What should happen?

**Comment by ckimes89** [comment]
Yeah this is pretty gross. I had hoped to be able to write the same expression to be used by every queryable provider, but it seems more and more like that may not be possible. Right now I see two options:

1. Require that in-memory queries be handled as post-resolved fields. This handles the case where EF querying works but in-memory querying doesn't by just changing the execution patterns entirely. I'm not really happy about this option for a few reasons, one is that it presents a potentially confusing API and the other is that we don't have a real guarantee of how other ORMs will react when encountering this problem.

2. Sniff for the queryable provider. LINQ to Objects, Entity Framework, and presumably other ORMs return a distinct provider from their IQueryable. We could detect known queryable providers and build the expression as necessary to support each provider. This allows us to insert null checking when querying in-memory and omit it when querying EF. We could also add other options as necessary where we find that ORMs differ.

**Comment by vicentedealencar** [comment]
Hi, I am failing to setup this lib seeing a `Object reference not set to an instance of an object.` and think this issue might be related.

> Inserting null checks can get really ugly and make the query expression incompatible with ORMs.

Is the [elvis operator](https://blogs.msdn.microsoft.com/jerrynixon/2014/02/26/at-last-c-is-getting-sometimes-called-the-safe-navigation-operator/) of any help here? I think it should work across ORMs... not sure

**Comment by ckimes89** [comment]
Unfortunately the null propagation operator is not allowed in expression trees. Since the basis of this library is built upon gluing expression trees together, that operator won't help with working around this problem. As a side note, it would be nice if the Expression API would be updated as features are added to the language.

If you are working against an in-memory context, then you can try using a manual expansion of the null propagation operator. For example, if you were seeing NREs while accessing fields on a null user, such as:

```
schema.AddType<User>()
    .AddField(u => u.Name)
```

then you could rewrite that with null safety as:

```
schema.AddType<User>()
    .AddField(u => u == null ? null : u.Name)
```

Are you executing against an in-memory context, or are you using an ORM?

**Comment by vicentedealencar** [comment]
I executed it (without sucess) against  Entity Framework over SQL Server. I will try again when the new version with these commits is pubblished. Thanks for your work

**Comment by ckimes89** [comment]
I just released version 0.2 on NuGet so you should be able to try out these changes. The API has changed quite a bit between these versions, so if you have any issues upgrading then make another issue and I'll help you there. I'm closing this one since I believe this particular problem is resolved now.
